contracts/redemption-rate-adapter/src/contract.rs

```diff
@@ -0,0 +1,149 @@
+use drop_helpers::answer::response;
+use drop_staking_base::error::redepmtion_rate_adapter::{ContractError, ContractResult};
+use drop_staking_base::msg::redepmtion_rate_adapter::{
+    Config, ExecuteMsg, InstantiateMsg, MigrateMsg, QueryMsg, RedemptionRateResponse,
+};
+
+use drop_staking_base::state::redemtion_rate_adapter::CONFIG;
+use neutron_sdk::bindings::msg::NeutronMsg;
+use neutron_sdk::bindings::query::NeutronQuery;
+
+const CONTRACT_NAME: &str = concat!("crates.io:drop-staking__", env!("CARGO_PKG_NAME"));
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn instantiate(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> ContractResult<Response<NeutronMsg>> {
+    cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    cw_ownable::initialize_owner(deps.storage, deps.api, Some(msg.owner.as_ref()))?;
+
+    let core_contract = deps.api.addr_validate(&msg.core_contract)?;
+    let config = &Config {
+        core_contract: core_contract.to_string(),
+        denom: msg.denom,
+    };
+    CONFIG.save(deps.storage, config)?;
+
+    Ok(response(
+        "instantiate",
+        CONTRACT_NAME,
+        [attr("core_contract", core_contract)],
+    ))
+}
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn query(deps: Deps<NeutronQuery>, env: Env, msg: QueryMsg) -> ContractResult<Binary> {
+    match msg {
+        QueryMsg::Ownership {} => Ok(to_json_binary(&get_ownership(deps.storage)?)?),
+        QueryMsg::Config {} => query_config(deps, env),
+        QueryMsg::RedemptionRate { denom, .. } => query_redemption_rate(deps, env, denom),
+    }
+}
+
+fn query_config(deps: Deps<NeutronQuery>, _env: Env) -> ContractResult<Binary> {
+    let config = CONFIG.load(deps.storage)?;
+    Ok(to_json_binary(&config)?)
+}
+
+fn query_redemption_rate(
+    deps: Deps<NeutronQuery>,
+    env: Env,
+    denom: String,
+) -> ContractResult<Binary> {
+    let config = CONFIG.load(deps.storage)?;
+    let core_contract = deps.api.addr_validate(&config.core_contract)?;
+    if denom != config.denom {
+        return Err(ContractError::InvalidDenom {});
+    }
+    let exchange_rate: Decimal = deps.querier.query_wasm_smart(
+        core_contract.to_string(),
+        &drop_staking_base::msg::core::QueryMsg::ExchangeRate {},
+    )?;
+    let core_state: drop_staking_base::state::core::ContractState = deps.querier.query_wasm_smart(
+        core_contract.to_string(),
+        &drop_staking_base::msg::core::QueryMsg::ContractState {},
+    )?;
+
+    let last_idle_raw = deps
+        .querier
+        .query_wasm_raw(core_contract, b"last_tick")?
+        .ok_or_else(|| {
+            ContractError::Std(cosmwasm_std::StdError::NotFound {
+                kind: "last_tick".to_string(),
+            })
+        })?;
+
+    let last_tick: u64 = from_json(last_idle_raw)?;
```

Comment:
We can skip these computations if core_state != idle

Reply:
last_tick always shows last update time of the exchange rate.

Reply:
The code below uses env.block.time if core_state != idle, that's why I am suggesting this optimisation.

Reply:
but real redemption rate updates during idle tick? Am I missing anything?

Reply:
when it's in Idle - it's not cached so we return current time, if it's not idle - we return cached time